### PR TITLE
submit

### PR DESCRIPTION
**Pull request by leqbill**

### CODE REVIEW

File: Problem1/src/problem1/Problem1.java [hunk 1]
```diff
@@ -77,7 +77,7 @@ public void looping5(int n) {
 
     /**
      * looping6
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(n^3)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: The inner while loop iterates n^3 times.  Thus, the total complexity is O(n^5)

File: Problem1/src/problem1/Problem1.java [hunk 2]
```diff
@@ -140,7 +140,7 @@ public int recursive3(int n) {
 
     /**
      * recursive4
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(n log n)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: For every non-base case call, 2 additional recursive calls are made (where n only decreases by 1).  Thus, the complexity is O(2^n).

File: Problem1/src/problem1/Problem1.java [hunk 3]
```diff
@@ -95,7 +95,7 @@ public void looping6(int n) {
 
     /**
      * recursive1
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(n)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: You got recursive1 correct.  recursive2 is also O(n).

File: Problem2/src/problem2/Problem2.java [hunk 4]
```diff
@@ -5,17 +5,60 @@
...
+        } else {
+            System.out.println(input + " is NOT divisible by three");
+        }
+    }
+
+    //takes in the int value and splits the number up into its digits
+    //sums the digits and then returns a boolean based on its divisibility by three
+    //@param int n the number passed through by the user
+    //@retunr boolean
+    public static boolean byThree(int n) {
```
Comment (cgallinaro) [inline Problem2/src/problem2/Problem2.java]: Since your method takes an integer, you can only evaluate numbers that will fit in an integer variable.  As a result, the large number from the problem description (49867285022465) will not work.

You could have accepted a string value instead.  This would have allowed for much larger values and it would have eliminated the need for the "numberAsString" variable below.

File: Problem2/src/problem2/Problem2.java [hunk 5]
```diff
@@ -5,17 +5,60 @@
...
+        for (int i = 0; i < count; i++) {
+            digits[i] = n % 10;
+            n = n / 10;
+        }
+
+        for (int i = 0; i < count; i++) {
+            num += digits[i];
+        }
+
+        if (num % 3 == 0) {
```
Comment (cgallinaro) [inline Problem2/src/problem2/Problem2.java]: Because you are using the mod (%) operator, you will notice that your recursive call is never called.  You should try running this with a break point on the "return byThree(num)" line.  I don't think it will ever get called.

File: Problem2/src/problem2/Problem2.java [hunk 6]
```diff
@@ -5,17 +5,60 @@
...
+            return true;
+        } else if (num % 3 !=0 && count == 1) {
+            return false;
+        } else {
+            return byThree(num);
+        }
+
```
Comment (cgallinaro) [inline Problem2/src/problem2/Problem2.java]: You didn't write any tests for this method!

File: Problem3/test/problem3/PersonTest.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Test of compareTo method, of class Person.
+     */
+    @Test
+    public void testCompareTo() {
+            person.add(new Person("1", "name1", 1));
+            person.add(new Person("2", "name2", 2));
+           
+            assertTrue(person.get(0).compareTo(person.get(1)) < 1);
+            assertFalse(person.get(1).compareTo(person.get(0)) < 1);
```
Comment (cgallinaro) [inline Problem3/test/problem3/PersonTest.java]: When writing tests for sorting algorithms, be sure to add tests for base cases:
- Empty array
- Array of one element

File: Problem3/test/problem3/Problem3Test.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    public void testInsertionSort() {
+        person.add(new Person("2", "name2", 2));
+        person.add(new Person("1", "name1", 1));
+
+        insertionSort(person);
+
+        assertTrue("name1".equals(person.get(0).getName()));
+        assertTrue("name2".equals(person.get(1).getName()));
+
+    }
```
Comment (cgallinaro) [inline Problem3/test/problem3/Problem3Test.java]: When writing tests for sorting algorithms, be sure to add tests for base cases:
- Empty array
- Array of one element

File: Problem4/src/problem4/Problem4.java [hunk 9]
```diff
@@ -5,8 +5,14 @@
...
 package problem4;
 
+import java.util.Random;
+
 /**
- *
+ *BIG O: O(n^2) because as the size of the array increases, the sort time is increased nonlinearly
+ * If an array were already sorted or almost sorted, then insertion sort works better
+ * because it only compares as many values as it needs to.
+ * selection sort works in a quadratic fashion and randomly sorted array will be sorted by selection sort
```
Comment (cgallinaro) [inline Problem4/src/problem4/Problem4.java]: Good job figuring out that this was selection sort.  A reverse sorted array would also be better sorted by selection sort.

File: Problem4/test/problem4/Problem4Test.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        Random ran = new Random();
+        int[] intArray = new int[100];
+       
+         for (int i=0; i<100; i++) {
+            intArray[i]= ran.nextInt(100);
+        }
+         
+         selectReplace(intArray);
+         
+        // for (int i = 0; i < intArray.length; i++) {
```
Comment (cgallinaro) [inline Problem4/test/problem4/Problem4Test.java]: This validation is not good enough to ensure that the array is fully sorted.

File: Problem4/test/problem4/Problem4Test.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+            assertFalse(intArray[0]>intArray[1]||intArray[0]==intArray[1]);
+            assertTrue(intArray[0]<intArray[1]||intArray[0]==intArray[1]);
+        //}
+    }
```
Comment (cgallinaro) [inline Problem4/test/problem4/Problem4Test.java]: Also, you should have test cases for the edge cases of your sorting algorithm:
- Empty array
- Array of one element

**Comment by cgallinaro** [comment]
Mark:

Problem 1: 3.5/5
Problem 2: 2/5
Problem 3: 5/5
Problem 4: 9/10

Total: 19.5/25
